### PR TITLE
[PATH]: post processor efficiency

### PR DESCRIPTION
**Pull request by J-Dunn**
streamline code, especially remove deep object references from inside nested iterations. This can be relevant on large output files such as produced by 3D tools.

https://forum.freecadweb.org/viewtopic.php?f=15&t=50583

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by sliptonic** [comment]
Both this PR and 3934 affect the same file and have a merge conflict.   Do you want to resolve and combine the two PRs into one or do you want me to merge one or the other?

**Comment by sliptonic** [comment]
Bump.   Can we please resolve the merge conflicts or remove the PR?

**Comment by J-Dunn** [comment]
I'm doing a fresh master build now to see where things are at.  Sadly, I seem to spend most of my free time battling with incompetent french ISPs who are continually sabotaging each other's services. That means my net access is very intermittent and the amount of free time left for coding is severely reduced.  I'll see what I can do.

File: src/Mod/Path/PathScripts/post/grbl_post.py [hunk 1]
```diff
@@ -285,52 +285,50 @@ def export(objectslist, filename, argstring):
...
-    gcode += linenumber() + line
+    gcode.append( linenumber() + line )
 
   if RETURN_TO:
-    gcode += linenumber() + "G0 X%s Y%s" % tuple(RETURN_TO)
+    gcode.append( linenumber() + "G0 X%s Y%s" % tuple(RETURN_TO) )
 
   # show the gCode result dialog
+  sep=''
+  final = sep.join(gcode)
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/post/grbl_post.py:324]: if you go this route, wouldn't it be more efficient to remove the `\n` from all lines and use that as the separator?
just a thought.

File: src/Mod/Path/PathScripts/post/grbl_post.py [hunk 2]
```diff
@@ -285,52 +285,50 @@ def export(objectslist, filename, argstring):
...
-    if result:
+    dia.editor.setText(final)
+    if dia.exec_():
       final = dia.editor.toPlainText()
-    else:
-      final = gcode
-  else:
-    final = gcode
-
+    dia.editor.setText("")
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/post/grbl_post.py:330]: ?

File: src/Mod/Path/PathScripts/post/grbl_post.py [hunk 3]
```diff
@@ -348,14 +346,11 @@ def linenumber():
...
+def format_outstring(strTable):
   global COMMAND_SPACE
-  # construct the line for the final output
-  s = ""
-  for w in strTbl:
-    s += w + COMMAND_SPACE
-  s = s.strip()
+  delim = COMMAND_SPACE
+  s = delim.join(strTable)
+  s+='\n'
```
Comment (mlampert) [inline src/Mod/Path/PathScripts/post/grbl_post.py:353]: this was explicitly removed previously - sure about this?

Comment (mlampert) [reply src/Mod/Path/PathScripts/post/grbl_post.py:353]: nm, I see what you're doing

**Review by mlampert — APPROVED** [review]

**Comment by J-Dunn** [comment]
The following PR deals with the issue at source, preventing the copy on read, rather than working around it. 

https://github.com/FreeCAD/FreeCAD/pull/4201

Once that is merges, I'll close this one. There is still work to be done on post-proc efficiency but it may be nice to restructure some of this and make it a class from which all the post procs can be derived and thus centralise much of the code which is now duped half a dozen times with minor mods.

**Comment by sliptonic** [comment]
This PR seems stalled since late December.  Please update status or close it if no longer relevant

**Comment by J-Dunn** [comment]
About half the speed increase here was from avoiding the copy-on-read in the underlying class of p.commands , that has been worked around in the class files. 

There is still a fair speed increase possible but it may be worth thinking about wrapping this with a class and minimise code repetition in the various post-proc variants. 

For now I'll just close this one since the main reason for this PR has been fixed at source.

**Pull request closed without merging**